File: src/api/metadata/orderbook.ts
import type { HttpClient } from '../../utils/http.js';
import type { Orderbook } from '../../types/index.js';

/**
 * API for retrieving orderbook snapshots.
 *
 * The orderbook shows current bid/ask prices and quantities for YES and NO
 * outcome tokens in a market.
 *
 * @example
 * ```typescript
 * const dflow = new DFlowClient();
 *
 * const orderbook = await dflow.orderbook.getOrderbook('BTCD-25DEC0313-T92749.99');
 * console.log(`YES Bid: ${orderbook.yesBid.price}, Ask: ${orderbook.yesAsk.price}`);
 * ```
 */
export class OrderbookAPI {
  constructor(private http: HttpClient) {}

  /**
   * Get the orderbook for a market by ticker.
   *
   * @param marketTicker - The market ticker
   * @returns Orderbook with YES/NO bid and ask prices and quantities
   *
   * @example
   * ```typescript
   * const orderbook = await dflow.orderbook.getOrderbook('BTCD-25DEC0313-T92749.99');
   * console.log(`YES: Bid ${orderbook.yesBid.price} / Ask ${orderbook.yesAsk.price}`);
   * console.log(`NO:  Bid ${orderbook.noBid.price} / Ask ${orderbook.noAsk.price}`);
   * ```
   */
  async getOrderbook(marketTicker: string): Promise<Orderbook> {
    return this.http.get<Orderbook>(`/orderbook/${marketTicker}`);
  }

  /**
   * Get the orderbook for a market by mint address.
   *
   * Alternative to {@link getOrderbook} when you have the mint address.
   *
   * @param mintAddress - The Solana mint address of the market's outcome token
   * @returns Orderbook with YES/NO bid and ask prices and quantities
   *
   * @example
   * ```typescript
   * const orderbook = await dflow.orderbook.getOrderbookByMint('EPjFWdd5...');
   * ```
   */
  async getOrderbookByMint(mintAddress: string): Promise<Orderbook> {
    return this.http.get<Orderbook>(`/orderbook/by-mint/${mintAddress}`);
  }
}

File: python/src/dflow/api/metadata/orderbook.py
"""Orderbook API for DFlow SDK."""

from dflow.types import Orderbook
from dflow.utils.http import HttpClient


class OrderbookAPI:
    """API for retrieving orderbook snapshots.

    The orderbook shows current bid/ask prices and quantities for YES and NO
    outcome tokens in a market.

    Example:
        >>> dflow = DFlowClient()
        >>> orderbook = dflow.orderbook.get_orderbook("BTCD-25DEC0313-T92749.99")
        >>> print(f"YES Bid: {orderbook.yes_bid[0].price}")
    """

    def __init__(self, http: HttpClient):
        self._http = http

    def get_orderbook(self, market_ticker: str) -> Orderbook:
        """Get the orderbook for a market by ticker.

        Args:
            market_ticker: The market ticker

        Returns:
            Orderbook with YES/NO bid and ask prices and quantities

        Example:
            >>> orderbook = dflow.orderbook.get_orderbook("BTCD-25DEC0313-T92749.99")
            >>> print(f"YES: Bid {orderbook.yes_bid[0].price} / Ask {orderbook.yes_ask[0].price}")
            >>> print(f"NO:  Bid {orderbook.no_bid[0].price} / Ask {orderbook.no_ask[0].price}")
        """
        data = self._http.get(f"/orderbook/{market_ticker}")
        return Orderbook.model_validate(data)

    def get_orderbook_by_mint(self, mint_address: str) -> Orderbook:
        """Get the orderbook for a market by mint address.

        Alternative to get_orderbook when you have the mint address.

        Args:
            mint_address: The Solana mint address of the market's outcome token

        Returns:
            Orderbook with YES/NO bid and ask prices and quantities
        """
        data = self._http.get(f"/orderbook/by-mint/{mint_address}")
        return Orderbook.model_validate(data)
